RecoParticleFlow/PFProducer/plugins/kdtrees/KDTreeLinkerTrackHcal.cc

```diff
@@ -239,26 +245,85 @@ void KDTreeLinkerTrackHcal::searchLinks() {
+        it->first->trackType(reco::PFBlockElement::T_FROM_V0) ||
+        it->first->trackType(reco::PFBlockElement::T_FROM_GAMMACONV)) {
+      for (BlockEltSet::iterator jt = it->second.begin(); jt != it->second.end(); ++jt) {
+        double clusterphi = (*jt)->clusterRef()->positionREP().phi();
+        double clustereta = (*jt)->clusterRef()->positionREP().eta();
+        multitracks.linkedClusters.push_back(std::make_pair(clusterphi, clustereta));
+      }
 
-      multitracks.linkedClusters.push_back(std::make_pair(trackphi, tracketa));
     }
+    //
+    // Store only the N closest HCAL links per track.
+    else {
+      reco::PFRecTrackRef trackref = it->first->trackRefPF();
+      const reco::PFTrajectoryPoint& tkAtHCALEnt = trackref->extrapolatedPoint(trajectoryLayerEntrance_);
+      const reco::PFCluster::REPPoint& tkreppos = tkAtHCALEnt.positionREP();
+      // Check exit point
+      double dHEta = 0.;
+      double dHPhi = 0.;
+      double dRHCALEx = 0.;
+      if (checkExit_) {
+        const reco::PFTrajectoryPoint& tkAtHCALEx = trackref->extrapolatedPoint(trajectoryLayerExit_);
+        dHEta = (tkAtHCALEx.positionREP().Eta() - tkAtHCALEnt.positionREP().Eta());
+        dHPhi = reco::deltaPhi(tkAtHCALEx.positionREP().Phi(), tkAtHCALEnt.positionREP().Phi());
+        dRHCALEx = tkAtHCALEx.position().R();
+      }
 
+      std::vector<double> vDist;
+      double dist(-1.0);
+
+      BlockEltSet hcalEltSet = it->second;
+
+      // Fill the vector of distances between HCAL clusters and the track
+      for (BlockEltSet::iterator jt = hcalEltSet.begin(); jt != hcalEltSet.end(); ++jt) {
+        double clusterphi = (*jt)->clusterRef()->positionREP().phi();
+        double clustereta = (*jt)->clusterRef()->positionREP().eta();
+
+        // when checkExit_ is false
+        if (!checkExit_) {
+          dist = LinkByRecHit::computeDist(clustereta, clusterphi, tkreppos.Eta(), tkreppos.Phi());
+        }
+        // when checkExit_ is true
+        else {
+          //special case ! A looper  can exit the barrel inwards and hit the endcap
+          //In this case calculate the distance based on the first crossing since
+          //the looper will probably never make it to the endcap
+          if (dRHCALEx < tkAtHCALEnt.position().R()) {
+            dist = LinkByRecHit::computeDist(clustereta, clusterphi, tkreppos.Eta(), tkreppos.Phi());
+            edm::LogWarning("KDTreeLinkerTrackHcal ")
+                << "Special case of linking with track hitting HCAL and looping back in the tracker ";
+          } else {
+            dist = LinkByRecHit::computeDist(
+                clustereta, clusterphi, tkreppos.Eta() + 0.1 * dHEta, tkreppos.Phi() + 0.1 * dHPhi);
+          }
+        }  // checkExit_
+        vDist.push_back(dist);
+      }  // loop over hcalEltSet
+
+      // Fill multitracks
+      for (auto i : sort_indexes(vDist)) {
+        BlockEltSet::iterator jt = std::next(hcalEltSet.begin(), i);
```

Comment:
just to note, this may have some runtime cost, as seeking to the n-th element in `std::set` is O(N), and you are doing so on each iteration. We can move std::set -> std::vector in the future in this code.

Reply:
Yeah, I imagine this could be improved but I wasn't sure how..